up-streamer/src/route.rs

```diff
@@ -0,0 +1,151 @@
+/// #         ) -> Result<String, UStatus> {
+/// #             println!("UPClientFoo: registering topic: {:?}", topic);
+/// #             let uuid = UUIDBuilder::new().build();
+/// #             Ok(uuid.to_string())
+/// #         }
+/// #
+/// #         async fn unregister_listener(&self, topic: UUri, listener: &str) -> Result<(), UStatus> {
+/// #             println!(
+/// #                 "UPClientFoo: unregistering topic: {topic:?} with listener string: {listener}"
+/// #             );
+/// #             Ok(())
+/// #         }
+/// #     }
+/// #
+/// #     impl UPClientFoo {
+/// #         pub fn new() -> Self {
+/// #             Self {}
+/// #         }
+/// #     }
+/// #     pub struct UTransportBuilderFoo;
+/// #     impl UTransportBuilder for UTransportBuilderFoo {
+/// #         fn build(&self) -> Result<Box<dyn UTransport>, UStatus> {
+/// #             let utransport_foo: Box<dyn UTransport> = Box::new(UPClientFoo::new());
+/// #             Ok(utransport_foo)
+/// #         }
+/// #     }
+/// #
+/// #     impl UTransportBuilderFoo {
+/// #         pub fn new() -> Self {
+/// #             Self {}
+/// #         }
+/// #     }
+/// #
+/// # }
+///
+/// let local_transport_router =
+///             UTransportRouter::start("FOO".to_string(),
+///                 foo_transport_builder::UTransportBuilderFoo::new(), 100, 200, 300);
+///         assert!(local_transport_router.is_ok());
+///         let local_transport_router_handle = Arc::new(local_transport_router.unwrap());
+///
+/// let authority_foo = UAuthority {
+///     name: Some("foo_name".to_string()).into(),
+///     number: Some(Number::Ip(vec![192, 168, 1, 100])),
+///     ..Default::default()
+/// };
+///
+/// let local_route = Route::new(&authority_foo, &local_transport_router_handle);
+/// ```
+#[derive(Clone)]
+pub struct Route {
+    authority: UAuthority,
+    transport_router_handle: Arc<UTransportRouterHandle>,
+}
+
+impl Route {
+    /// Builds a [`Route`] based on [`UAuthority`][up_rust::UAuthority] and [`UTransportRouterHandle`]
+    ///
+    /// # Parameters
+    ///
+    /// * `authority` - The [`UAuthority`][up_rust::UAuthority] to associate with this [`Route`]
+    /// * `transport_router_handle` - a [`UTransportRouterHandle`][crate::UTransportRouterHandle]
+    ///                               obtained by calling
+    ///                               [`UTransport::start()`][crate::UTransportRouter::start]
+    ///                               with the appropriate arguments.
+    ///
+    /// For more information see the example under [`Route`]
+    pub fn new(
+        authority: &UAuthority,
+        transport_router_handle: &Arc<UTransportRouterHandle>,
+    ) -> Self {
+        // Try to initiate logging.
+        // Required in case of dynamic lib, otherwise no logs.
+        // But cannot be done twice in case of static link.
+        let _ = env_logger::try_init();
+        if log_enabled!(Level::Debug) {
```

Comment:
I think the log_enabled! check is redundant. My understanding is that debug! will already have an internal check to check if the debug level is enabled.

Reply:
Yep you're right. If not set to debug, this wouldn't get output.

However, I want to avoid doing any calculations / formatting strings which if I understand correctly would be done still.

But maybe I went overkill will my usage of gating these for cases where there's not expensive calculations.

I'll review my usage.

Reply:
lazy evaluation of your log strings could solve that problem for you.
For example in Java you have String Format that is a function that will do the string manipulation only if called

Reply:
Reviewed the `log` crate to make sure this was doing what I expected.

Here's the [`debug!()` logging macro](https://github.com/rust-lang/log/blob/28d734b0c4f8c36ce0070682ce14891d602ca97a/src/macros.rs#L153-L161).

It seems like the [log level is checked](https://github.com/rust-lang/log/blob/28d734b0c4f8c36ce0070682ce14891d602ca97a/src/macros.rs#L35) _first_ before any of the arguments are evaluated or string formatting occurs (i.e. the correct thing :slightly_smiling_face:)

Removed most `log_enabled!()` calls except for a couple where a hash needs to be computed before calling `debug!()`.